[scroll-anchoring] Keyboard scrolling into a rubberband leaves the page stuck at a bad scroll offset

nmoucht opened the pull request:
#### 50c9dcae73447ac8e649a07dd246b227fc3dfd65
<pre>
[scroll-anchoring] Keyboard scrolling into a rubberband leaves the page stuck at a bad scroll offset
<a href="https://bugs.webkit.org/show_bug.cgi?id=267061">https://bugs.webkit.org/show_bug.cgi?id=267061</a>
<a href="https://rdar.apple.com/120053910">rdar://120053910</a>

Reviewed by Simon Fraser.

After <a href="https://commits.webkit.org/271957@main">https://commits.webkit.org/271957@main</a> we disabled scroll anchoring adjustments while
m_owningScrollableArea.isRubberBandInProgress() is true. However, this is not true during a
keyboard scroll. A better solution to this is to invalidate the scroll anchoring element in
ScrollAnimator::setCurrentPosition itself. Normally scroll position changes here would be sent
down into the scrollable area via setScrollPositionFromAnimation, however while rubberbanding
notify is NotifyScrollableArea::No, so we don&apos;t propagate the scroll position change to the
scrollable area. Now that we properly notfiy the scroll anchoring controller of the scroll
position change while rubberbanding, we can remove the old solution.

* Source/WebCore/page/scrolling/ScrollAnchoringController.cpp:
(WebCore::ScrollAnchoringController::adjustScrollPositionForAnchoring):
* Source/WebCore/platform/ScrollAnimator.cpp:
(WebCore::notifyScrollAnchoringControllerOfScroll):
(WebCore::ScrollAnimator::setCurrentPosition):

Canonical link: <a href="https://commits.webkit.org/272657@main">https://commits.webkit.org/272657@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7d19359f21bfe3a32c4dd21990d569b28ed01790

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/32564 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/11312 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/34412 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/35130 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/29436 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/33430 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/13661 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/8506 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/28961 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/32998 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/9535 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/29102 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/8308 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/8446 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/29043 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/36466 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/29598 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/29465 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/34572 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/8573 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/6526 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/32431 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/10248 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/28785 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/9185 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/4202 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/9166 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr reviewed (approved):
Can we write a test?

nmoucht committed:
[scroll-anchoring] Keyboard scrolling into a rubberband leaves the page stuck at a bad scroll offset
https://bugs.webkit.org/show_bug.cgi?id=267061
rdar://120053910

Reviewed by Simon Fraser.

After https://commits.webkit.org/271957@main we disabled scroll anchoring adjustments while
m_owningScrollableArea.isRubberBandInProgress() is true. However, this is not true during a
keyboard scroll. A better solution to this is to invalidate the scroll anchoring element in
ScrollAnimator::setCurrentPosition itself. Normally scroll position changes here would be sent
down into the scrollable area via setScrollPositionFromAnimation, however while rubberbanding
notify is NotifyScrollableArea::No, so we don't propagate the scroll position change to the
scrollable area. Now that we properly notfiy the scroll anchoring controller of the scroll
position change while rubberbanding, we can remove the old solution.

* Source/WebCore/page/scrolling/ScrollAnchoringController.cpp:
(WebCore::ScrollAnchoringController::adjustScrollPositionForAnchoring):
* Source/WebCore/platform/ScrollAnimator.cpp:
(WebCore::notifyScrollAnchoringControllerOfScroll):
(WebCore::ScrollAnimator::setCurrentPosition):

Canonical link: https://commits.webkit.org/272657@main
